tools/src/icon4py/tools/py2fgen/_codegen.py

```diff
@@ -0,0 +1,536 @@
+# ICON4Py - ICON inspired code in Python and GT4Py
+#
+# Copyright (c) 2022-2024, ETH Zurich and MeteoSwiss
+# All rights reserved.
+#
+# Please, refer to the LICENSE file in the root directory.
+# SPDX-License-Identifier: BSD-3-Clause
+
+from typing import Any, Optional, Sequence, TypeGuard, Union
+
+from gt4py.eve import Node, codegen, datamodels
+from gt4py.eve.codegen import JinjaTemplate as as_jinja, TemplatedGenerator
+
+from icon4py.tools.py2fgen import _definitions, _utils
+
+
+CFFI_DECORATOR = "@ffi.def_extern()"
+
+BUILTIN_TO_ISO_C_TYPE: dict[_definitions.ScalarKind, str] = {
+    _definitions.FLOAT64: "real(c_double)",
+    _definitions.FLOAT32: "real(c_float)",
+    _definitions.BOOL: "logical(c_int)",
+    _definitions.INT32: "integer(c_int)",
+    _definitions.INT64: "integer(c_long)",
+}
+BUILTIN_TO_CPP_TYPE: dict[_definitions.ScalarKind, str] = {
+    _definitions.FLOAT64: "double",
+    _definitions.FLOAT32: "float",
+    _definitions.BOOL: "int",
+    _definitions.INT32: "int",
+    _definitions.INT64: "long",
+}
+BUILTIN_TO_NUMPY_TYPE: dict[_definitions.ScalarKind, str] = {
+    _definitions.FLOAT64: "xp.float64",
+    _definitions.FLOAT32: "xp.float32",
+    _definitions.BOOL: "xp.int32",
+    _definitions.INT32: "xp.int32",
+    _definitions.INT64: "xp.int64",
+}
+
+
+def is_array(param: _definitions.ParamDescriptor) -> TypeGuard[_definitions.ArrayParamDescriptor]:
+    return isinstance(param, _definitions.ArrayParamDescriptor)
+
+
+class Func(Node):
+    name: str
+    args: dict[str, Union[_definitions.ArrayParamDescriptor, _definitions.ScalarParamDescriptor]]
+    rendered_params: str = datamodels.field(init=False)  # TODO remove
+
+    def __post_init__(self, *args: Any, **kwargs: Any) -> None:
+        params = []
+        for name, param in self.args.items():
+            params.append(name)
+            if is_array(param):
+                for i in range(param.rank):
+                    params.append(_size_arg_name(name, i))
+        params.append("on_gpu")
+        self.rendered_params = ", ".join(params)
+
+
+class CffiPlugin(Node):
```

Comment:
```suggestion
class CFFIPlugin(Node):
```
maybe replace `plugin` by `Library`
```suggestion
class BindingsLibrary(Node):
```